uportal-war/src/main/java/org/jasig/portal/layout/dlm/remoting/UpdatePreferencesServlet.java

```diff
@@ -863,7 +882,133 @@ public ModelAndView addTab(HttpServletRequest request, HttpServletResponse respo
+        newFolder.setAddChildAllowed(true);
+        newFolder.setFolderType(IUserLayoutFolderDescription.REGULAR_TYPE);
+        
+        // Update the attributes based on the supplied JSON (optional request body name-value pairs)
+        if (attributes != null && !attributes.isEmpty()) {
+            setObjectAttributes(newFolder, request, attributes);
+        }
+
+        ulm.addNode(newFolder, targetId, siblingId);
+        final Locale locale = RequestContextUtils.getLocale(request);
+        
+        try {
+            ulm.saveUserLayout();
+        } catch (Exception e) {
+            log.warn("Error saving layout", e);
+            return new ModelAndView("jsonView", Collections.singletonMap("error", getMessage("error.persisting.layout.change.folder", "Unable to add a new folder", locale)));
+        }        
+
+        Map<String, Object> model = new HashMap<>();
+        model.put("response", getMessage("success.add.folder", "Added a new folder", locale));
+        model.put("folderId", newFolder.getId());
+        model.put("immutable", newFolder.isImmutable());
+        return new ModelAndView("jsonView", model);
+    }
+    
+    /**
+     * Attempt to map the attribute values to the given object.
+     * @param node
+     * @param request
+     * @param attributes
+     */
+    private void setObjectAttributes(IUserLayoutNodeDescription node, HttpServletRequest request, Map<String, Map<String, String>> attributes) {
+        // Attempt to set the object attributes
+        for(String name : attributes.get("attributes").keySet()) {
+            try {
+              BeanUtils.setProperty(node, name, attributes.get(name));
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                log.warn("Unable to set attribute: " + name + "on object of type: " + node.getType());
+            }
+        }
+        
+        // Set the structure-attributes, whatever they may be
+        Map<String, String> structureAttributes = attributes.get("structureAttributes");
+        if (structureAttributes != null) {
+            for(String name : structureAttributes.keySet()) {
+                this.stylesheetUserPreferencesService.setLayoutAttribute(request,
+                                                                         PreferencesScope.STRUCTURE, 
+                                                                         node.getId(), 
+                                                                         name, 
+                                                                         structureAttributes.get(name));
+            }
+        }
+    }
+
+    /**
+     * Update the attributes for the node. Unrecognized attributes will log a warning, but are otherwise ignored.
+     * 
+     * @param request
+     * @param response
+     * @param targetId
+     * @param attributes - parse the JSON name-value pairs in the body as the attributes of the element.
```

Comment:
Please clarify as you did in addFolder that this is both attributes and structureAttributes and what the request body should look like (and mime type if application/json is needed to get the object mapping to work correctly)